// code derived from https://github.com/keystonejs/keystone/

// https://github.com/keystonejs/keystone/blob/master/LICENSE
/*
(The MIT License)

Copyright (c) 2016 Jed Watson

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

diff -Naur -r v0.3.15/node_modules/keystone/lib/session.js v0.3.16/node_modules/keystone/lib/session.js
--- a/lib/session.js	2015-08-18 08:09:25.000000000 +0300
+++ b/lib/session.js	2015-12-04 04:14:53.000000000 +0200
@@ -84,6 +84,10 @@
 	}
 	var User = keystone.list(keystone.get('user model'));
 	if ('string' === typeof lookup.email && 'string' === typeof lookup.password) {
+		// ensure that it is an email, we don't want people being able to sign in by just using "\." and a haphazardly correct password.
+		if (!utils.isEmail(lookup.email)) {
+			return onFail(new Error('Incorrect email or password'));
+		}
 		// create regex for email lookup with special characters escaped
 		var emailRegExp = new RegExp('^' + utils.escapeRegExp(lookup.email) + '$', 'i');
 		// match email address and password
